### PR TITLE
Feat-directional-fee-hook

### PR DESCRIPTION
**Pull request by mkflow27**
This pr adds the directional fee hook implementation.

### CODE REVIEW

File: typescript/src/hooks/directionalFeeHook.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    ): { success: boolean; dynamicSwapFee: bigint } {
+
+        const lastBalancesLiveScaled18 = hookState.balancesLiveScaled18;
+
+        const tokenInIndex = hookState.tokens.indexOf(params.tokenIn);
+        const tokenOutIndex = hookState.tokens.indexOf(params.tokenOut);
+
+        const calculatedSwapFeePercentage = this.calculateExpectedSwapFeePercentage(
+            lastBalancesLiveScaled18,
+            params.amountRaw, // does this need to be scaled?
```
Comment (johngrantuk) [inline typescript/src/hooks/directionalFeeHook.ts:35]: Should be able to just replicate the SC approach. So if its scaled in SC it needs scaled here. Let me know if this doesn't work.

File: typescript/test/hooks/directionalFee.test.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    amountRaw: swapAmountRaw,
+    tokenIn: poolTokens[0], 
+    tokenOut: poolTokens[1],
+};
+
+describe('hook - directionalFee', () => {
+    const vault = new Vault();
+    const hook = new DirectionalFeeHook();
+
+    test('computes directional fee', () => {
```
Comment (johngrantuk) [inline typescript/test/hooks/directionalFee.test.ts:60]: This test is basically the same as `it uses dynamic swap fee with high enough swap amount` right? So maybe not neccessary?

File: typescript/test/hooks/directionalFee.test.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        }
+        const { success, dynamicSwapFee } = hook.onComputeDynamicSwapFee(newSwapParams, staticSwapFeePercentage, hookState);
+        expect(success).toBe(true);
+        expect(dynamicSwapFee).toEqual(staticSwapFeePercentage);
+    })
+    test('it uses dynamic swap fee with high enough swap amount', () => {
+        // swap amount is big enough to trigger dynamic swap fee
+        const { success, dynamicSwapFee } = hook.onComputeDynamicSwapFee(swapParams, staticSwapFeePercentage, hookState);
+        expect(success).toBe(true);
+        expect(dynamicSwapFee).toBeGreaterThan(staticSwapFeePercentage);
```
Comment (johngrantuk) [inline typescript/test/hooks/directionalFee.test.ts:80]: Would be better to test that it equals a known amount I think.

File: typescript/test/hooks/directionalFee.test.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    })
+
+    test('directionalFee higher than static swap fee', () => {
```
Comment (johngrantuk) [inline typescript/test/hooks/directionalFee.test.ts:83]: Not really sure this is a neccessary test? Its not really related to the directionalFee hook?

File: typescript/test/hooks/directionalFee.test.ts
Comment (johngrantuk) [inline typescript/test/hooks/directionalFee.test.ts:1]: Would be good to add a couple of tests that use `vault.swap` to confirm the full flow is covered.

File: typescript/src/hooks/directionalFeeHook.ts (hunk 1 above)
Comment (mkflow27) [reply typescript/src/hooks/directionalFeeHook.ts]: Scaling has been added.

File: typescript/test/hooks/directionalFee.test.ts (hunk 3 above)
Comment (mkflow27) [reply typescript/test/hooks/directionalFee.test.ts:106]: The tests now include testing for known amounts.

**Pull request merged**